bzzeth/bzzeth.go

```diff
@@ -99,6 +119,215 @@ func (b *BzzEth) handleMsgFromSwarmNode(p *Peer) func(context.Context, interface
+
+	// collect the hashes of block headers we want
+	var hashes [][]byte
+	for i, y := range yes {
+		// ignore hashes already present in localstore
+		if y {
+			continue
+		}
+
+		// collect hash based on proximity
+		vhash := addresses[i]
+		if wantHeaderFunc(vhash, b.kad) {
+			hashes = append(hashes, vhash)
+		} else {
+			p.logger.Trace("ignoring header. Not in proximity ", "Address", hex.EncodeToString(addresses[i]))
+		}
+	}
+
+	// request them from the offering peer and deliver in a channel
+	deliveries := make(chan []byte)
+	req, err := p.getBlockHeaders(ctx, hashes, deliveries)
+	if err != nil {
+		p.logger.Error("Error sending GetBlockHeader message", "Reason", err)
+		return
+	}
+	defer req.cancel()
+
+	// this loop blocks until all delivered or context done
+	// only needed to log results
+	deliveredCnt := 0
+	for {
+		select {
+		case hash, ok := <-deliveries:
+			if !ok {
+				p.logger.Debug("bzzeth.handleNewBlockHeaders", "hash", hex.EncodeToString(hash), "delivered", deliveredCnt)
+				return
+			}
+			deliveredCnt++
+			p.logger.Trace("bzzeth.handleNewBlockHeaders", "hash", hex.EncodeToString(hash), "delivered", deliveredCnt)
+			if deliveredCnt == len(req.hashes) {
+				p.logger.Debug("Delivered all headers", "count", deliveredCnt)
+				finishDeliveryFunc(req.hashes)
+				return
+			}
+		case <-ctx.Done():
+			p.logger.Debug("bzzeth.handleNewBlockHeaders", "delivered", deliveredCnt, "err", err)
+			return
+		}
+	}
+}
+
+// wantHeaderFunc is used to determine if we need a particular header offered as latest
+// by an eth fullnode
+// tests reassign this to control
+var wantHeaderFunc = wantHeader
+
+// wantHeader returns true iff the hash argument falls in the NN of kademlia
+func wantHeader(hash []byte, kad *network.Kademlia) bool {
+	return chunk.Proximity(kad.BaseAddr(), hash) >= kad.NeighbourhoodDepth()
+}
+
+// finishStorageFunc is used to determine if all the requested headers are stored
+// this is used in testing if the headers are indeed stored in the localstore
+var finishStorageFunc = finishStorage
+
+func finishStorage(chunks []chunk.Chunk) {
+	for _, c := range chunks {
+		log.Trace("Header stored", "Address", c.Address().Hex())
+	}
+}
+
+// finishDeliveryFunc is used to determine if all the requested headers are delivered
+// this is used in testing .. otherwise it just logs trace
+var finishDeliveryFunc = finishDelivery
+
+func finishDelivery(hashes map[string]bool) {
+	for addr := range hashes {
+		log.Trace("Header delivered", "Address", addr)
+	}
+}
+
+// handleBlockHeaders handles block headers message
+func (b *BzzEth) handleBlockHeaders(ctx context.Context, p *Peer, msg *BlockHeaders) {
+	p.logger.Debug("bzzeth.handleBlockHeaders", "id", msg.ID)
+
+	// retrieve the request for this id
+	req, ok := p.requests.get(msg.ID)
+	if !ok {
+		p.logger.Warn("bzzeth.handleBlockHeaders: nonexisting request id", "id", msg.ID)
+		p.Drop()
+		return
+	}
+
+	// convert rlp.RawValue to bytes
+	var headers [][]byte
```

Comment:
`headers := make([][]byte, len(msg.Headers))` please,
then 
```for i, h ... {
   headers[i]=h
}
```
also, not sure but `h` might need to be shadowed

Reply:
h shadowed by what? i dont see any other h in that function.

Reply:
i meant it might need shadowing inside the loop. for now just make the above change with:
```go
headers := make([][]byte, len(msg.Headers)) 
for i, h ... {
   headers[i]=h
}
```